File: client/app/utils/tsp.ts
class CompareFirstElement {
    public compare(left: [number, [number, number]], right: [number, [number, number]]): number {
      return left[0] - right[0];
    }
  }
  
  class Node {
    num: number = 0;
    neighbours: number[] = [];
  
    constructor() {
      this.neighbours = [];
    }
  }
  
  export function solveTSP(adj: number[][]): number[] {
    const visited: boolean[] = new Array(adj.length).fill(false);
    const pq: [number, [number, number]][] = [];
    const compareFirstElement = new CompareFirstElement();
  
    pq.push([0, [0, 0]]);
  
    const nodes: Node[] = new Array(adj.length).fill(null).map(() => new Node());
  
    while (pq.length > 0) {
      const [cost, [parent, current]] = pq.shift()!;
      if (visited[current]) continue;
  
      nodes[parent].neighbours.push(current);
      nodes[current].neighbours.push(parent);
  
      visited[current] = true;
      for (let i = 0; i < adj.length; i++) {
        if (adj[current][i] !== 0 && !visited[i]) {
          pq.push([adj[current][i], [current, i]]);
        }
      }
      pq.sort(compareFirstElement.compare);
    }
  
    const order: number[] = [];
    const q: number[] = [];
    q.push(0);
    visited.fill(false);
  
    while (q.length > 0) {
      const u = q.shift()!;
      if (visited[u]) continue;
      visited[u] = true;
      order.push(u);
  
      for (const v of nodes[u].neighbours) {
        if (!visited[v]) {
          q.push(v);
        }
      }
    }
  
    return order;
  }

File: server/tsp.cpp
// emcc -Oz tsp.cpp -lembind -o tsp.js -s EXPORT_ES6=1 -s MODULARIZE=1 -s EXPORT_NAME=loadWASM -s SINGLE_FILE=1
#include <iostream>
#include <vector>
#include <utility>
#include <queue>
#include <emscripten/bind.h>
#include <emscripten/emscripten.h>


using std::vector;
using std::pair;
using std::priority_queue;
using std::queue;
using std::make_pair;

struct CompareFirstElement {
    bool operator()(const pair<double, pair<int, int> >& left, const pair<double, pair<int, int> >& right) const {
        return left.first > right.first;
    }
};

struct Node{
    int num;
    vector<int> neighbours;
};

vector<int> EMSCRIPTEN_KEEPALIVE Solve(const vector<vector<double>> adj){
      vector<bool> visited(adj.size(), false);
      priority_queue<pair<double,pair<int,int> >, vector<pair<double,pair<int,int> > >, CompareFirstElement> pq;
      pq.push(make_pair(0, make_pair(0, 0)));

      vector<Node> nodes(adj.size());

      while(!pq.empty()){
          pair<double,pair<int, int> > p = pq.top();
          pq.pop();
          if(visited[p.second.second]) continue;

          nodes[p.second.first].neighbours.push_back(p.second.second);
          nodes[p.second.second].neighbours.push_back(p.second.first);

          visited[p.second.second] = true;
          for(int i = 0; i < adj.size(); i++){
              if(adj[p.second.second][i] != 0 && !visited[i]){
                  pq.push(make_pair(adj[p.second.second][i], make_pair(p.second.second, i)));
              }
          }
      }

      vector<int> order;
      queue<int> q;
      q.push(0);
      visited.assign(adj.size(), false);
      while(!q.empty()){
          int u = q.front();
          q.pop();
          if(visited[u]) continue;
          visited[u] = true;
          order.push_back(u);
          for(int v : nodes[u].neighbours){
              if(!visited[v]){
                  q.push(v);
              }
          }
      }
      return order;
}

int EMSCRIPTEN_KEEPALIVE add(int x, int y) {
    return x + y;
}


EMSCRIPTEN_BINDINGS(tsp_module) {
    emscripten::register_vector<int>("IntList");
    emscripten::register_vector<double>("DoubleList");
    emscripten::register_vector<vector<double>>("DoubleListList");
    emscripten::function("Solve", &Solve);
    emscripten::function("add", &add);
}
